src/Microsoft.Health.Fhir.TemplateManagement/Overlay/OverlayOperator.cs

```diff
@@ -0,0 +1,206 @@
+using SharpCompress.Common;
+using SharpCompress.Writers;
+using SharpCompress.Writers.Tar;
+
+namespace Microsoft.Health.Fhir.TemplateManagement.Overlay
+{
+    public class OverlayOperator : IOverlayOperator
+    {
+        public OCIFileLayer ExtractOCIFileLayer(OCIArtifactLayer artifactLayer)
+        {
+            if (artifactLayer == null)
+            {
+                return null;
+            }
+
+            var artifacts = StreamUtility.DecompressTarGzStream(new MemoryStream(artifactLayer.Content));
+            var metaBytes = artifacts.GetValueOrDefault(Constants.OverlayMetadataFile);
+            OverlayMetadata metaJson = null;
+            if (metaBytes != null)
+            {
+                try
+                {
+                    metaJson = JsonConvert.DeserializeObject<OverlayMetadata>(Encoding.UTF8.GetString(metaBytes));
+                }
+                catch (Exception ex)
+                {
+                    throw new OverlayException(TemplateManagementErrorCode.OverlayMetaJsonNotFound, "overlayMeta.json file invalid", ex);
+                }
+            }
+
+            artifacts.Remove(Constants.OverlayMetadataFile);
+            return new OCIFileLayer() { Content = artifactLayer.Content, FileContent = artifacts, FileName = artifactLayer.FileName, SequenceNumber = metaJson == null ? -1 : metaJson.SequenceNumber };
+        }
+
+        public List<OCIFileLayer> ExtractOCIFileLayers(List<OCIArtifactLayer> artifactLayers)
+        {
+            if (artifactLayers == null)
+            {
+                return null;
+            }
+
+            var result = new List<OCIFileLayer>();
+            foreach (var layer in artifactLayers)
+            {
+                result.Add(ExtractOCIFileLayer(layer));
+            }
+
+            return result;
+        }
+
+        public List<OCIFileLayer> SortOCIFileLayersBySequenceNumber(List<OCIFileLayer> fileLayers)
+        {
+            if (fileLayers == null)
+            {
+                return null;
+            }
+
+            return fileLayers.OrderBy(layer => layer.SequenceNumber).ToList();
+        }
+
+        public OCIFileLayer MergeOCIFileLayers(List<OCIFileLayer> sortedLayers)
+        {
+            if (sortedLayers == null)
+            {
+                return null;
+            }
+
+            List<string> removedFiles = new List<string>();
+            List<string> currentLayerRemovedFiles = new List<string>();
+            Dictionary<string, byte[]> mergedFiles = new Dictionary<string, byte[]> { };
+
+            for (var sequenceNumber = sortedLayers.Count - 1; sequenceNumber >= 0; sequenceNumber--)
+            {
+                var oneLayer = sortedLayers[sequenceNumber];
+                var fileContent = oneLayer.FileContent;
+                foreach (var oneFile in fileContent)
+                {
+                    if (oneFile.Value == null)
+                    {
+                        currentLayerRemovedFiles.Add(oneFile.Key);
```

Comment:
Can we directly use `removedFiles` here instead of the intermediate `currentLayerRemovedFiles`?

Reply:
Seems we can remove the intermediate for the current design.